File: server/dataObject/mainDataModel.js
const { json } = require("express");

class MainDataModel {
  constructor({
    lan,
    id,
    c_date,
    title,
    des,
    f_count,
    s_count,
    active,
    a_date
  }) {
    if (!title instanceof MainDataTitleModel)
      throw "type err - MainDataTitleModel";
    if (!des instanceof MainDataDesModel) throw "type err - MainDataDesModel";
    this.lan = lan;
    this.id = id;
    this.c_date = c_date;
    this.title = title;
    this.des = des;
    this.f_count = f_count;
    this.s_count = s_count;
    this.active = active;
    this.a_date = a_date;
  }
}

MainDataModel.prototype.json = json => {
  return new MainDataModel({
    title: MainDataTitleModel.prototype.json(json.title),
    des: MainDataDesModel.prototype.json(json.des),
    lan: json.lan,
    id: json.id,
    c_date: json.c_date,
    f_count: json.f_count,
    s_count: json.s_count,
    active: json.active,
    a_date: json.a_date
  });
};

class MainDataTitleModel {
  constructor({ txt, image, subTitle }) {
    this.txt = txt;
    this.image = image;
    this.subTitle = subTitle;
  }
}

MainDataTitleModel.prototype.json = json => {
  return new MainDataTitleModel(json);
};

class MainDataDesModel {
  constructor({ des, images, type }) {
    this.des = des;
    this.type = type; // 단순 텍스트인지, 이미지인지, 영상인지
    if (!Array.isArray(images)) throw "type err Array<String>";
    console.log(images);
    images.forEach(e => {
      if (typeof e !== "string") throw "type err - ele String";
    });
    // + validation e - https://
    this.image = images;
  }
}

MainDataDesModel.prototype.json = json => {
  return new MainDataDesModel(json);
};

module.exports = { MainDataModel, MainDataTitleModel, MainDataDesModel };

File: lib/project/models/mainDataObject/mainDataModel.dart
import 'package:blossom/project/models/responseModel/resBodyModel.dart';
import 'package:blossom/project/modules/LanParserModule.dart';


/// 클래스 쓸 때 왠만하면 다 final을 붙이는게 좋다.
/// 클래스 extends가 안되게
final class MainDataModel implements ResBodyModel{
  final Lan lan; /// @
  final String id;
  final String c_date;
  final MainDataTitleModel title;
  final MainDataDesModel des;
  final String f_count;
  final String s_count;
  final String active;
  final String a_date;
  const MainDataModel({
    required this.lan,
    required this.id,
    required this.c_date,
    required this.title,
    required this.des,
    required this.f_count,
    required this.s_count,
    required this.active,
    required this.a_date
  });
  factory MainDataModel.json(Map<String, dynamic> body){
    return MainDataModel(
        lan: LanParserModule.stringToLen(body["lan"]),
        id: body['id'],
        c_date: body['c_date'],
        title: MainDataTitleModel.json(body['title']), /// @객체
        des: MainDataDesModel.json(body['des']), /// @객체
        f_count: body['f_count'],
        s_count: body['s_count'],
        active: body['active'],
        a_date: body['a_date']
    );
  }
}


final class MainDataTitleModel{
  final String txt;
  final String image;
  final String subTitle;
  const MainDataTitleModel({required this.txt, required this.image, required this.subTitle});
  MainDataTitleModel.json(Map<String, dynamic> body):
      txt = body['txt'],
      image = body['image'],
      subTitle = body['subTitle'];
}

enum DesType{
  Txt,
  Image,
  Video,
  TnI,
  TnV,
  Err
}

final class MainDataDesParser{
  static DesType parser(String type){
  if(type =="Txt" || type == "txt") return DesType.Txt;
  if(type =="video") return DesType.Video;
  if(type =="image") return DesType.Image;
  if(type.toUpperCase() == "TNV") return DesType.TnV;
  if(type.toLowerCase() == "tni") return DesType.TnI;
  return DesType.Err;
  }

}


final class MainDataDesModel{
  final String des;
  final List<String> images;
  final DesType type;
  const MainDataDesModel({required this.des, required this.images, required this.type});

  /// initializer도 자체 메서드도 못쓴다.(아직 객체가 안만들어졌다)
  /// 근데 factory도 자체 메서드는 못쓴다.
  factory MainDataDesModel.json(Map<String, dynamic> body){
    return MainDataDesModel(
        des: body['des'],
        images: body['images'],
        type: MainDataDesParser.parser(body['type'])
    );

}




